### PR TITLE
Complete the implementation of the mesh orientation algorithm in 3d.

### PR DESCRIPTION
**Pull request by bangerth**
This patch completes the re-implementation of the edge orientation algorithm started in #3166 and collected in #3235.

I have extensively tested this patch. It passes the entire testsuite (with the exception of the issue noted in #3261 that I think is unrelated). I have also verified that the old and new code give identical results *in the result of this algorithm* for all testcases where it is run, not just in the output of the testcase as a whole. In other words, I've run both old code and new code from the same function and assert'ed that the cell orientations are the same.

Finally, I have done some timing by analyzing all tests that run through this code. Among the debug mode tests, 112 tests run through the 3d code here. Only about ten take more than 0.01 seconds in the code in question here, and in these cases, the new code is between 1.5 and 3 times slower. Because this may be of interested for someone looking at this later on, the longest running test is `tests/grid/grid_in_3d_02.debug` which takes 0.88 seconds before and 1.79 seconds after this change. (That's in debug mode; in release mode -- see below -- the numbers are 0.170 and 0.173 seconds.) This tests reads in a mesh that describes a Boeing 747 and has 56,040 vertices and 51,823 cells -- not the usual mesh we read in, and still it only takes barely more than one second to orient. Surely this algorithm will not be the bottleneck in anything one does on these meshes. So I'm not overly concerned by the run-time increase in debug mode.

I've also run all `.*.release` tests and there the picture is different: The ratio of run times for the new over the old code ranges from 0.3 to 3, i.e., the new code sometimes is significantly faster, but for the majority of testcases the ratio is almost exactly 1. This is particularly true for all tests that spend more than 0.1 seconds in this algorithm. In other words, for all cases where it matters, the old and the new code are exactly equally fast.

What this means in practice is that the difference in run times for debug mode is entirely due to the fact that the new code contains a number of assertions; in contrast, the old code had essentially none. One could remove these assertions again now that we know that the algorithm works, but I think that's not worth the hassle and I like the added safety of leaving them in.

I'm going to submit this patch as is because it switches the existing implementation over to the dimension-independent one. I have one clean-up that I still want to take care of (unify the various, dimension-independent implementations of the `GridReordering<dim,spacedim>::reorder_cells()` functions), but will do so in a follow-up patch to ease review of the current functionality.

### CODE REVIEW

**Comment by bangerth** [comment]
@mikeando -- FYI

**Comment by bangerth** [comment]
Oh, and two other things I want to do after this is merged: (i) make the namespace for the implementation anonymous; (ii) go through the documentation and see what needs to be updated. But as I said, that's after this is merged.

File: source/grid/grid_reordering.cc [hunk 1]
```diff
@@ -251,6 +269,19 @@ namespace internal
     };
 
 
+
+    /**
+     * A class that represents all of the cells adjacent to a given edge.
+     * This class corresponds to the 3d case where each edge can have an
+     * arbitrary number of adjacent cells. We represent this as a
+     * std::set<AdjacentCell>, from which class the current one is
```
Comment (drwells) [inline source/grid/grid_reordering.cc]: typo: 'we represent this as a `std::set`' should be 'we represent this as a `std::vector`

**Comment by drwells** [comment]
This looks good at a glance. I like that this removes about 1100 lines of code.

Did `std::set` for `EdgeDeltaSet` end up having any performance impact? If you did not measure it then I might out of curiousity.

**Comment by bangerth** [comment]
I was content for the moment that the performance matched the previous code and had it in my list if things to try in the future. But I will gladly let you beat me to it.

File: source/grid/grid_reordering.cc (hunk 1 above)
Comment (bangerth) [inline source/grid/grid_reordering.cc]: Fixed.

**Comment by drwells** [comment]
There is a very slight performance increase (on the order of 10 ms for `grid_in_3d_02` out of about 150ms) in `GridReordering<3>::reorder_cells` if I use the proposed 'private inheritance from `std::vector`' instead of 'public inheritance from `std::set`'. Since the call to `read_ucd` takes about a second this is not an important performance increase (only about 1%). I suppose what I am trying to say is "you are right".

I am surprised that `std::set` does so well here!

**Comment by bangerth** [comment]
Keep your changes and submit them as a follow up patch once this one is merged.

**Comment by tjhei** [comment]
> 51,823 cells -- not the usual mesh we read in, and still it only takes barely more than one second to orient. Surely this algorithm will not be the bottleneck in anything one does on these meshes. So I'm not overly concerned by the run-time increase in debug mode.

I think we should run the new algorithm on a significant larger mesh to make sure it is scaling similarly. I know @kronbichler has some 1million+ cell meshes (if I remember correctly) we could possibly test.

**Comment by bangerth** [comment]
We could do this. But since the ratio of run times for the old and new algorithms is pretty much exactly one, and given that we run the algorithm from 1 to 50k cells, we know for sure that the complexities of the two algorithms are the same. 

I suspect that you're not going to learn anything you didn't already know by testing things on even larger meshes.

**Comment by mikeando** [comment]
I do know the original version was tweaked due to some unexpected dominance of certain terms at large mesh sizes. (In particular, it was tested on some large 2D and 3D meshes that only just fit in memory...) Now the meshes I tested it on may have been "overly large" at the time, but they're probably not any more... and if I'd not addressed the performance on large meshes - the whole thing would have needed an overhaul well before now.

So my suggestion is that we should test on the really large meshes, if it's not too difficult, just to ensure that we are getting exactly the scaling we expect - then we're less likely to have to revisit this later. Hopefully, we learn nothing new...

**Comment by bangerth** [comment]
Here's a test: I'm taking a 3d hyper cube, and then I'm going to repeatedly (i) refine it, (ii) output the mesh in ucd format, (iii) read this ucd file again and (iv) send it through the edge orienter. The timings I get are as follows for debug mode:
```
n_cells=8, time=0
n_cells=64, time=0
n_cells=512, time=0.004
n_cells=4096, time=0.039994
n_cells=32768, time=0.383942
n_cells=262144, time=3.71744
n_cells=2097152, time=34.9387
```
That's a time of `O(N^1.08)`. One could ask where the extra 0.08 comes from, but I think that is maybe not worth investigating. I'm going to note that `N` here is the number of cells, but the algorithm is really only linear in the number of edges.

In release mode I get this:
```
n_cells=8, time=0
n_cells=64, time=0
n_cells=512, time=0
n_cells=4096, time=0.006999
n_cells=32768, time=0.06799
n_cells=262144, time=0.705892
n_cells=2097152, time=8.63469
```
That's about `O(N^1.12)`. Again, I think it's close enough to `O(N)` and with such a small constant that I'm not going to worry too much about where exactly I'm losing the 0.12 in convergence order.

**Comment by tjhei** [comment]
> n_cells=2097152, time=8.63469

Is the timing only for the orientation? That looks a bit slower than I expected. Do you know how it is for the old code?

> Again, I think it's close enough to O(N) and with such a small constant that I'm not going to worry too much about where exactly I'm losing the 0.12 in convergence order.

Agreed and I think that confirms the new algorithm works well.

**Comment by bangerth** [comment]
Yes, the timing is for the reorientation only. It also fits with the previously quoted number of 0.1 seconds for 50k cells.

As for the overall time: for 2M cells, you have to touch 6M edges. (Each cell has 12 edges, and on a uniformly refined mesh, each edge is shared between 4 cells.) So 8 seconds for 6 million edges means roughly one micro-second or 2,000 clock cycles to orient it. That doesn't seem vastly out of place. In either case, it's going to be at least 2 orders of magnitude faster than anything else you might probably want to do on that mesh.

**Comment by drwells** [comment]
Since this passes the test suite and looks reasonable (I must admit that I don't fully understand the algorithm) I think that it is okay to merge this.

**Comment by drwells** [comment]
Clarification: unless someone has further requests for changes I will merge this in the next day or so.

**Pull request merged**